File: frontend/src/types/threads.ts
/**
 * Thread Library type definitions.
 *
 * TypeScript interfaces for the thread generation and specification APIs.
 */

/** Thread standard family identifier. */
export type ThreadFamilyId =
  | 'iso_metric'
  | 'unc'
  | 'unf'
  | 'npt'
  | 'bspp'
  | 'bspt'
  | 'acme'
  | 'trapezoidal';

/** Thread direction (internal or external). */
export type ThreadType = 'internal' | 'external';

/** Thread helix direction. */
export type ThreadHand = 'right' | 'left';

/** Pitch series for metric threads. */
export type PitchSeries = 'coarse' | 'fine' | 'superfine';

/** 3D printing process. */
export type PrintProcess = 'fdm' | 'sla' | 'sls' | 'mjf';

/** Thread tolerance class. */
export type ToleranceClass = 'tight' | 'standard' | 'loose';

/** Print feasibility rating. */
export type PrintFeasibility = 'excellent' | 'good' | 'marginal' | 'not_recommended';

/** Thread family information. */
export interface ThreadFamily {
  family: ThreadFamilyId;
  name: string;
  description: string;
  standard_ref: string;
  size_count: number;
}

/** Thread family list response. */
export interface ThreadFamilyListResponse {
  families: ThreadFamily[];
  total: number;
}

/** Complete thread specification. */
export interface ThreadSpec {
  family: string;
  size: string;
  pitch_mm: number;
  form: string;
  pitch_series: string | null;
  major_diameter: number;
  pitch_diameter_ext: number;
  minor_diameter_ext: number;
  major_diameter_int: number;
  pitch_diameter_int: number;
  minor_diameter_int: number;
  profile_angle_deg: number;
  taper_per_mm: number;
  tap_drill_mm: number;
  clearance_hole_close_mm: number;
  clearance_hole_medium_mm: number;
  clearance_hole_free_mm: number;
  tpi: number | null;
  nominal_size_inch: string | null;
  engagement_length_mm: number;
  standard_ref: string;
  notes: string;
}

/** Thread size list response. */
export interface ThreadSizeListResponse {
  family: string;
  sizes: string[];
  total: number;
  pitch_series: string | null;
}

/** Tap drill info response. */
export interface TapDrillInfo {
  family: string;
  size: string;
  tap_drill_mm: number;
  clearance_hole_close_mm: number;
  clearance_hole_medium_mm: number;
  clearance_hole_free_mm: number;
}

/** Thread generation request. */
export interface ThreadGenerateRequest {
  family: string;
  size: string;
  thread_type?: ThreadType;
  length_mm: number;
  hand?: ThreadHand;
  pitch_series?: PitchSeries | null;
  custom_pitch_mm?: number | null;
  custom_diameter_mm?: number | null;
  add_chamfer?: boolean;
}

/** Thread generation response. */
export interface ThreadGenerateResponse {
  success: boolean;
  metadata: Record<string, unknown>;
  generation_time_ms: number;
  estimated_face_count: number;
  message: string;
}

/** Print recommendation response. */
export interface PrintRecommendation {
  family: string;
  size: string;
  feasibility: PrintFeasibility;
  min_recommended_size: string;
  recommended_tolerance: string;
  clearance_mm: number;
  notes: string[];
  orientation_advice: string;
  estimated_strength_pct: number;
}

/** Print-optimized generation request. */
export interface PrintOptimizedGenerateRequest {
  family: string;
  size: string;
  thread_type?: ThreadType;
  length_mm: number;
  process?: PrintProcess;
  tolerance_class?: ToleranceClass;
  nozzle_diameter_mm?: number;
  layer_height_mm?: number;
  use_flat_bottom?: boolean;
  custom_clearance_mm?: number | null;
  pitch_series?: PitchSeries | null;
  hand?: ThreadHand;
  add_chamfer?: boolean;
}

/** Print-optimized generation response. */
export interface PrintOptimizedGenerateResponse {
  success: boolean;
  feasibility: string;
  adjustments_applied: Record<string, number>;
  recommendation: PrintRecommendation;
  generation_result: ThreadGenerateResponse | null;
  message: string;
}

File: backend/app/schemas/threads.py
"""
Thread library Pydantic schemas.

Request and response models for the thread standards API (v2).
Covers thread family listing, spec lookup, tap drill info,
thread generation, and print-optimised generation.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

# =============================================================================
# Response Models — Thread Standards
# =============================================================================


class ThreadFamilyResponse(BaseModel):
    """Response for a single thread family."""

    model_config = ConfigDict(from_attributes=True)

    family: str = Field(description="Family identifier")
    name: str = Field(description="Human-readable name")
    description: str = Field(description="Family description")
    standard_ref: str = Field(description="Standard reference")
    size_count: int = Field(description="Number of sizes available")


class ThreadFamilyListResponse(BaseModel):
    """Response listing all thread families."""

    families: list[ThreadFamilyResponse]
    total: int


class ThreadSpecResponse(BaseModel):
    """Complete thread specification response."""

    model_config = ConfigDict(from_attributes=True)

    family: str
    size: str
    pitch_mm: float
    form: str
    pitch_series: str | None = None
    major_diameter: float
    pitch_diameter_ext: float
    minor_diameter_ext: float
    major_diameter_int: float
    pitch_diameter_int: float
    minor_diameter_int: float
    profile_angle_deg: float
    taper_per_mm: float
    tap_drill_mm: float
    clearance_hole_close_mm: float
    clearance_hole_medium_mm: float
    clearance_hole_free_mm: float
    tpi: float | None = None
    nominal_size_inch: str | None = None
    engagement_length_mm: float
    standard_ref: str
    notes: str


class ThreadSizeListResponse(BaseModel):
    """Response listing available sizes for a family."""

    family: str
    sizes: list[str]
    total: int
    pitch_series: str | None = None


class TapDrillResponse(BaseModel):
    """Tap drill and clearance hole information."""

    family: str
    size: str
    tap_drill_mm: float
    clearance_hole_close_mm: float
    clearance_hole_medium_mm: float
    clearance_hole_free_mm: float


# =============================================================================
# Request / Response Models — Thread Generation
# =============================================================================


class ThreadGenerateRequest(BaseModel):
    """Request to generate thread geometry."""

    family: str = Field(description="Thread family")
    size: str = Field(description="Thread size")
    thread_type: str = Field(
        default="external",
        description="internal or external",
    )
    length_mm: float = Field(
        gt=0,
        le=200,
        description="Thread length in mm",
    )
    hand: str = Field(default="right", description="right or left")
    pitch_series: str | None = Field(
        default=None,
        description="coarse or fine (metric)",
    )
    custom_pitch_mm: float | None = Field(
        default=None,
        gt=0,
        description="Override pitch",
    )
    custom_diameter_mm: float | None = Field(
        default=None,
        gt=0,
        description="Override diameter",
    )
    add_chamfer: bool = Field(
        default=True,
        description="Add lead-in chamfer",
    )


class ThreadGenerateResponse(BaseModel):
    """Response from thread generation."""

    success: bool
    metadata: dict
    generation_time_ms: int
    estimated_face_count: int
    message: str


# =============================================================================
# Request / Response Models — Print-Optimised Generation
# =============================================================================


class PrintOptimizedGenerateRequest(BaseModel):
    """Request to generate print-optimized thread."""

    family: str = Field(description="Thread family")
    size: str = Field(description="Thread size")
    thread_type: str = Field(
        default="external",
        description="internal or external",
    )
    length_mm: float = Field(
        gt=0,
        le=200,
        description="Thread length in mm",
    )
    process: str = Field(
        default="fdm",
        description="fdm, sla, sls, or mjf",
    )
    tolerance_class: str = Field(
        default="standard",
        description="tight, standard, or loose",
    )
    nozzle_diameter_mm: float = Field(
        default=0.4,
        gt=0,
        le=2.0,
        description="FDM nozzle diameter in mm",
    )
    layer_height_mm: float = Field(
        default=0.2,
        gt=0,
        le=1.0,
        description="Print layer height in mm",
    )
    use_flat_bottom: bool = False
    custom_clearance_mm: float | None = Field(
        default=None,
        ge=0,
        description="Override clearance value",
    )
    pitch_series: str | None = None
    hand: str = "right"
    add_chamfer: bool = True


class PrintRecommendationResponse(BaseModel):
    """Print feasibility recommendation."""

    family: str
    size: str
    feasibility: str
    min_recommended_size: str
    recommended_tolerance: str
    clearance_mm: float
    notes: list[str]
    orientation_advice: str
    estimated_strength_pct: float


class PrintOptimizedGenerateResponse(BaseModel):
    """Response from print-optimized thread generation."""

    success: bool
    feasibility: str
    adjustments_applied: dict[str, float]
    recommendation: PrintRecommendationResponse
    generation_result: ThreadGenerateResponse | None = None
    message: str
